sycl/doc/design/SYCL-RTC.md

```diff
@@ -0,0 +1,312 @@
+
+We extracted common boilerplate code to configure a
+[`CompilerInstance`](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/clang/include/clang/Frontend/CompilerInstance.h#L81)
+in the
+[`RTCActionBase`](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp#L176)
+class. Inside  the `GetLLVMModuleAction`, we instantiate and execute the
+aforementioned `EmitLLVMAction`, and, in case the translation was successful,
+[obtains ownership](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp#L255)
+of the constructed `llvm::Module` from it.
+
+Finally, the call to
+[`Action.takeModule()`](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp#L442)
+transfers ownership again to the caller of `compileDeviceCode`. Note that this
+simple mechanism works because we know that there is only a single compilation
+happening for every instance of the `ClangTool` and hence our
+`GetLLVMModuleAction` class.
+
+## Caching
+
+The implementation optionally uses the runtime's
+[persistent cache](https://intel.github.io/llvm/design/KernelProgramCache.html#persistent-cache)
+to elide recurring invocations of the frontend, which we observed to be the most
+expensive (in terms of runtime overhead) phase of our compilation pipeline.
+
+### Overall design
+
+We cache only the frontend invocation, meaning that after a successful
+translation, we store the LLVM IR module obtained via LibTooling on disk in the
+Bitcode format using built-in utilities. In case of a cache hit in a later
+runtime compilation, we load the module from disk and feed it into the device
+linking phase. The rationale for this design was that were no utilities to save
+and restore the linked and post-processed device images to disk at the time (the
+[SYCLBIN](https://intel.github.io/llvm/design/SYCLBINDesign.html) infrastructure
+was added later), and caching these steps would have resulted only in marginal
+further runtime savings.
+
+### Cache key considerations
+
+The main challenge is to define a robust cache key. Because code compiled via
+SYCL-RTC can `#include` header files defined via the `include_files` property as
+well as from the filesystem, e.g. `sycl.hpp` from the DPC++ installation or user
+libraries, it is not sufficient to look only at the source string. In order to
+make the cache as conservative as possible (cache collisions are unlikely but
+mathematically possible), we decided to compute a hash value of the
+*preprocessed* source string, i.e. with all `#include` directives resolved. We
+additionally compute a hash value of the rendered command-line arguments, and
+append it to the hash of the preprocessed source to obtain the final cache key. 
```

Comment:
I'd expect clang's PCH to have to deal with the same issue. Have you looked into re-using that infrastructure?

Reply:
No, while we did some non-SYCL (due to lack of support at the time) experiments with PCHs, we have not specifically looked into reusing any hashing infra from there.